### PR TITLE
Remove memory side effect from hal.interface.binding.subspan

### PR DESCRIPTION
**Pull request by qedawkins**
Enables CSE + DCE on duplicate/unused interface bindings as a followup to #12471. This removes the test for interface bindings in RemoveDeadMemAllocs as it is no longer marked as an alloc.

benchmarks: x86_64, cuda

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
Could you push an empty commit for the benchmarks to run.

**Review by benvanik — APPROVED** [review]

File: compiler/src/iree/compiler/Dialect/HAL/IR/test/interface_ops.mlir
Comment (antiagainst) [inline compiler/src/iree/compiler/Dialect/HAL/IR/test/interface_ops.mlir]: Nit: could drop the `, offsets = ..` part given that's not relevant to what we are checking here.

**Review by antiagainst — APPROVED** [review]
Nice, thanks for fixing this! Fresh eyes on long historical issue really helps. :)

**Comment by qedawkins** [comment]
Because collaborators cannot land PRs would it help to add a note in the commit message indicating that it needs a maintainer/the reviewer to land it? If there is some reason to wait on landing something that is fine too, I just don't want to endlessly ping people about it :)

(also I cannot land this :P)

**Comment by antiagainst** [comment]
> Because collaborators cannot land PRs would it help to add a note in the commit message indicating that it needs a maintainer/the reviewer to land it? If there is some reason to wait on landing something that is fine too, I just don't want to endlessly ping people about it :)
> 
> (also I cannot land this :P)

@GMNGeoffrey for some GitHub magic. :)

**Comment by GMNGeoffrey** [comment]
Adding a merge bot has been on my list of things that I want to do, wouldn't be that hard, and I've never had time for. In the meantime, I'd suggest noting in a comment that the PR is ready to land from the author's perspective

**Comment by MaheshRavishankar** [comment]
@qedawkins does this remove the `hal.interface.binding.subspan` ops even if it is used in `memref.assume_alignment` operations?

**Comment by qedawkins** [comment]
> @qedawkins does this remove the `hal.interface.binding.subspan` ops even if it is used in `memref.assume_alignment` operations?

As in there are two identical `hal.interface.binding.subspan` with different values for `memref.assume_alignment` (wait does this even make sense)? I think it would but I can check.

**Comment by qedawkins** [comment]
@MaheshRavishankar this survives if this is what you were asking
```
func.func @interface_subspan_cse() {
  %c0 = arith.constant 0 : index

  %0 = hal.interface.binding.subspan set(0) binding(2) type(storage_buffer) offset(%c0) : memref<2x32x16384xf32>
  memref.assume_alignment %0, 64 : memref<2x32x16384xf32>
  return
}
```
But if you were to assume two different alignments on otherwise the same binding then they would get cse'd.

**Comment by MaheshRavishankar** [comment]
Well, there was logic to remove even those subspans in the path that handled all allocation-type objects. We need similar logic somewhere

**Comment by qedawkins** [comment]
I see https://github.com/openxla/iree/blob/1148f720be7e267f248e034b3cfb488633884980/compiler/src/iree/compiler/Codegen/Common/CleanupBufferAllocViewPass.cpp#L105? The lazy solution would be to just treat bindings as an allocate here but I doubt that's a good solution. I see that's basically in the test that was removed here though. If it's an immediate problem I can revert and find a different place to put this cleanup functionality.

**Comment by MaheshRavishankar** [comment]
No that's ok. I might need to add somethings over here. I am refactoring a few things, I have a place to put these.

**Pull request merged**
